File: MetroBill/VB/InvoiceControl.vb
Imports System.ComponentModel
Imports System.Text

Namespace MetroBill
	Partial Public Class InvoiceControl
		Inherits UserControl
		Public Sub New()
			InitializeComponent()
		End Sub

		Private _Commands As MetroBillCommands
		''' <summary>
		''' Gets or sets the commands associated with the control.
		''' </summary>
		Public Property Commands() As MetroBillCommands
			Get
				Return _Commands
			End Get
			Set(ByVal value As MetroBillCommands)
				If value IsNot _Commands Then
					Dim oldValue As MetroBillCommands = _Commands
					_Commands = value
					OnCommandsChanged(oldValue, value)
				End If
			End Set
		End Property
		''' <summary>
		''' Called when Commands property has changed.
		''' </summary>
		''' <param name="oldValue">Old property value</param>
		''' <param name="newValue">New property value</param>
		Protected Overridable Sub OnCommandsChanged(ByVal oldValue As MetroBillCommands, ByVal newValue As MetroBillCommands)
			If newValue IsNot Nothing Then
				saveButton.Command = newValue.InvoiceCommands.Save
				cancelButton.Command = newValue.InvoiceCommands.Cancel
			Else
				saveButton.Command = Nothing
				cancelButton.Command = Nothing
			End If
		End Sub
	End Class
End Namespace

File: MetroBill/C#/InvoiceControl.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Text;
using System.Windows.Forms;

namespace MetroBill
{
    public partial class InvoiceControl : UserControl
    {
        public InvoiceControl()
        {
            InitializeComponent();
        }

        private MetroBillCommands _Commands;
        /// <summary>
        /// Gets or sets the commands associated with the control.
        /// </summary>
        public MetroBillCommands Commands
        {
            get { return _Commands; }
            set
            {
                if (value != _Commands)
                {
                    MetroBillCommands oldValue = _Commands;
                    _Commands = value;
                    OnCommandsChanged(oldValue, value);
                }
            }
        }
        /// <summary>
        /// Called when Commands property has changed.
        /// </summary>
        /// <param name="oldValue">Old property value</param>
        /// <param name="newValue">New property value</param>
        protected virtual void OnCommandsChanged(MetroBillCommands oldValue, MetroBillCommands newValue)
        {
            if (newValue != null)
            {
                saveButton.Command = newValue.InvoiceCommands.Save;
                cancelButton.Command = newValue.InvoiceCommands.Cancel;
            }
            else
            {
                saveButton.Command = null;
                cancelButton.Command = null;
            }
        }
    }
}
